src/rascal/models/pseudo_points.hh

```diff
@@ -0,0 +1,419 @@
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifndef SRC_RASCAL_MODELS_PSEUDO_POINTS_HH_
+#define SRC_RASCAL_MODELS_PSEUDO_POINTS_HH_
+
+#include "rascal/math/utils.hh"
+#include "rascal/representations/calculator_base.hh"
+#include "rascal/structure_managers/property_block_sparse.hh"
+#include "rascal/structure_managers/structure_manager_collection.hh"
+#include "rascal/utils/json_io.hh"
+
+namespace rascal {
+
+  namespace internal {}  // namespace internal
+
+  /**
+   * Set of pseudo points associated with a representation stored in BlockSparse
+   * format, e.g. SphericalInvariants. The number of pseudo points is often
+   * refered as $M$ and note that they might be the representation of actual
+   * atomic environments or completely artificial.
+   *
+   * Pseudo points are useful to build sparse kernel models such as Subset of
+   * Regressors. This class is tailored for building property models that
+   * depend on the type of the central atom.
+   */
+  template <class Calculator>
+  class PseudoPointsBlockSparse {
+   public:
+    using Key_t = typename CalculatorBase::Key_t;
+    using Data_t = std::map<int, std::map<Key_t, std::vector<double>>>;
+    using Indices_t = std::map<int, std::map<Key_t, std::vector<size_t>>>;
+    using Counters_t = std::map<int, size_t>;
+
+    template <class StructureManager>
+    using Property_t =
+        typename Calculator::template Property_t<StructureManager>;
+    template <class StructureManager>
+    using PropertyGradient_t =
+        typename Calculator::template PropertyGradient_t<StructureManager>;
+
+    /**
+     * Container for the actual data. By construction it gathers pseudo points
+     * in a particular manner: [sp][key][feature_idx] ; where sp is the type
+     * of the central atom to which these features correspond to, key is a
+     * valid key taken from the BlockSparse original container and feature_idx
+     * is a linear index containing all the features associated to several
+     * sparse points.
+     * This storage keeps the minimal amount of data.
+     */
+    Data_t values{};
+    /**
+     * Container for the sparse points indices arranged like:
+     * [sp][key][sparse_point_idx] ; where sp and key are like in values and
+     * sparse_point_idx are the indices of the sparse points in the particular
+     * [sp][key] bin.
+     */
+    Indices_t indices{};
+    //! counts number of sparse points for each [sp]
+    Counters_t counters{};
+    //! size of one feature block in [sp][key]
+    size_t inner_size{0};
+    //! list of possible center species for accessing [sp]
+    std::set<int> center_species{};
+    //! list of possible keys for accessing [key]
+    std::set<Key_t> keys{};
+
+    PseudoPointsBlockSparse() {
+      // there is less than 130 elemtents
+      for (int sp{1}; sp < 130; ++sp) {
```

Comment:
Do I understand correctly that `PseudoPoints` is always `PseudoPointsBlockSparse`.
Is there any other way to look at this? Could there possibly be a non-sparse `PseudoPoints` class?
If so, should there be a base class and this a specialization of it?

Reply:
`PseudoPointsBlockSparse` is only compatible with `PseudoPointsBlockSparse` and yes there should be a `PseudoPointsDense` implementation but it is outside of this PR. I thought of making these with a base class but it felt really pointless here since it would have no implementation functions or virtual ones since all member functions are templated. 
Would you prefer to have an empty base class anyway ?

Reply:
We should probably leave it at that for now before we did not talk about the whole PseudoPoints business. Maybe some stuff will clear up.